numpy/core/src/umath/ufunc_object.c

```diff
@@ -125,6 +125,64 @@ PyUFunc_clearfperr()
     npy_clear_floatstatus();
 }
 
+/*
+ * This function analyzes the input arguments and determines an appropriate
+ * method (__array_prepare__ or __array_wrap__) function to call, taking it
+ * from the input with the highest priority.
+ */
+static PyObject*
+_find_array_method(PyObject *args, int nin, PyObject *method_name)
+{
+    int i, np;
+    PyObject *obj;
+    PyObject *with_method[NPY_MAXARGS], *methods[NPY_MAXARGS];
+    PyObject *method = NULL;
+
+    np = 0;
+    for (i = 0; i < nin; i++) {
+        obj = PyTuple_GET_ITEM(args, i);
+        if (PyArray_CheckExact(obj) || PyArray_IsAnyScalar(obj)) {
+            continue;
+        }
+        method = PyObject_GetAttr(obj, method_name);
```

Comment:
Should we use `PyArray_LookupSpecial_OnInstance` here instead?

Reply:
Yes, I guess that would make sense, in principle, though here we're passing on a pre-calculated python string object rather than a `char` - which should be faster. Also, it might be better to go directly to `PyArray_LookupSpecial`...  So, maybe leave for another PR?